packages/jazz-tools/src/tools/coValues/request.ts

```diff
@@ -0,0 +1,617 @@
+  CoMapInitZod,
+  CoMapSchema,
+  CoValue,
+  CoValueClass,
+  Group,
+  Loaded,
+  ResolveQuery,
+  ResolveQueryStrict,
+  Simplify,
+  anySchemaToCoSchema,
+  coMapDefiner,
+  exportCoValue,
+  importContentPieces,
+  loadCoValue,
+} from "../internal.js";
+import { Account } from "./account.js";
+
+type MessageShape = Record<string, z.core.$ZodType | AnyCoSchema>;
+
+type RequestSchemaDefinition<
+  S extends MessageShape,
+  R extends ResolveQuery<CoMapSchema<S>> = true,
+> =
+  | S
+  | {
+      schema: S;
+      resolve?: R;
+    };
+
+/**
+ * Configuration options for defining HTTP request/response schemas in Jazz.
+ *
+ * This interface defines the structure for creating typed HTTP routes with
+ * request and response validation using CoMap schemas.
+ *
+ * @template RequestShape - The shape of the request message schema (must extend MessageShape)
+ * @template RequestResolve - The resolve query type for the request CoMap schema
+ * @template ResponseShape - The shape of the response message schema (must extend MessageShape)
+ * @template ResponseResolve - The resolve query type for the response CoMap schema
+ */
+interface RequestOptions<
+  RequestShape extends MessageShape,
+  RequestResolve extends ResolveQuery<CoMapSchema<RequestShape>>,
+  ResponseShape extends MessageShape,
+  ResponseResolve extends ResolveQuery<CoMapSchema<ResponseShape>>,
+> {
+  /**
+   * The URL endpoint for the HTTP route.
+   * This is used by the client to send requests to the server.
+   */
+  url: string;
+
+  /**
+   * The id of the worker Account or Group.
+   */
+  workerId: string;
+
+  /**
+   * Schema definition for the request payload.
+   * Can be either a direct schema object or an object with schema and optional resolve properties.
+   * The schema defines the structure and validation rules for incoming requests.
+   */
+  request: RequestSchemaDefinition<
+    RequestShape,
+    ResolveQueryStrict<CoMapSchema<RequestShape>, RequestResolve>
+  >;
+
+  /**
+   * Schema definition for the response payload.
+   * Can be either a direct schema object or an object with schema and optional resolve properties.
+   * The schema defines the structure and validation rules for outgoing responses.
+   */
+  response: RequestSchemaDefinition<
+    ResponseShape,
+    ResolveQueryStrict<CoMapSchema<ResponseShape>, ResponseResolve>
+  >;
+}
+
+type AsNullablePayload<T extends MessageShape> = T extends Record<string, never>
+  ? undefined
+  : never;
+type MessageValuePayload<T extends MessageShape> =
+  | Simplify<CoMapInitZod<T>>
+  | AsNullablePayload<T>;
+
+function createMessageEnvelope<S extends MessageShape>(
+  schema: AnyCoMapSchema,
+  value: MessageValuePayload<S>,
+  owner: Account,
+  sharedWith: Account | Group,
+): Loaded<CoMapSchema<S>> {
+  const group = Group.create({ owner });
+
+  group.addMember(sharedWith, "reader");
+
+  // @ts-expect-error - AnyCoMapSchema doesn't have static methods
+  return schema.create(value ?? {}, group);
```

Comment:
Why not make the input `CoMapSchema` instead?

Reply:
Because it would make the validation fail on other sides, and even after fixing this I would need to do type casting to make the `value ?? {}` work (Typescript can't know that `{}` is ok in this context)

At some point I decided to remove all the bloat I did write just to make TS happy and bailed out the type validation in some places.